### PR TITLE
*: setup e2e tests on Travis CI

### PR DESCRIPTION
**Pull request by LiliC**
This PR enables running end-to-end tests on Travis CI. This sets up Kubernetes with minikube on Travis CI, with the help of flag `--vm-driver=none`.

Closes https://github.com/kinvolk/habitat-operator/issues/76.

### CODE REVIEW

**Comment by LiliC** [comment]
FYI: We do have one test which is flaky, one every 10 times maybe, but this is due to Habitat itself, rather then the operator. The problem being that Supervisor runs into a problem and keeps retrying, so our test fail because the pod is already setup and ready to be tested, but application itself is not up and running.

I will open an issue for this random failure on Habitat, as we discussed on slack. Also there was a suggestion to maybe replace the test image with a Rails application example, currently we use the Node.js application. Will open issue for that as well.

File: Makefile [hunk 1]
```diff
@@ -9,7 +9,7 @@ build:
 	go build -i github.com/kinvolk/habitat-operator/cmd/habitat-operator
 
 linux:
-	env GOOS=linux go build github.com/kinvolk/habitat-operator/cmd/habitat-operator
+	GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -i --ldflags="-s" -o habitat-operator github.com/kinvolk/habitat-operator/cmd/habitat-operator
```
Comment (asymmetric) [inline Makefile]: Why are all of these changes necessary?

File: test/e2e/operator_test.go [hunk 2]
```diff
@@ -131,6 +133,11 @@ func TestServiceGroupInitialConfig(t *testing.T) {
 	if msg != actualMsg {
 		t.Fatalf("Initial Configuration failed. Msg did not match the one expected. Expected: %s got: %s", msg, actualMsg)
 	}
+
+	// Delete Service so it doesn't interfere with other tests.
```
Comment (asymmetric) [inline test/e2e/operator_test.go]: How can it interfere?

File: travis-setup.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
```
Comment (asymmetric) [inline travis-setup.sh]: I wonder if we could find a better place for this. Maybe under `test/`?

File: travis-setup.sh [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+# Setup for Travis CI.
+
+# Download kubectl, which is a requirement for using minikube.
+curl -Lo kubectl https://storage.googleapis.com/kubernetes-release/release/v1.7.0/bin/linux/amd64/kubectl && chmod +x kubectl && sudo mv kubectl /usr/local/bin/
+
+# Download minikube.
+curl -Lo minikube https://storage.googleapis.com/minikube/releases/latest/minikube-linux-amd64 && chmod +x minikube && sudo mv minikube /usr/local/bin/
+
+echo "starting minikube..."
+sudo minikube start --vm-driver=none --kubernetes-version=v1.7.0
```
Comment (asymmetric) [inline travis-setup.sh]: Is `sudo` required here?

File: test/e2e/operator_test.go [hunk 5]
```diff
@@ -108,6 +109,7 @@ func TestServiceGroupInitialConfig(t *testing.T) {
 	if err := framework.WaitForEndpoints(sgName); err != nil {
 		t.Fatal(err)
 	}
+	time.Sleep(1 * time.Minute)
```
Comment (asymmetric) [inline test/e2e/operator_test.go]: Can we make a constant out of tihs?

File: travis-setup.sh [hunk 6]
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
+# Setup for Travis CI.
+
+# Download kubectl, which is a requirement for using minikube.
+curl -Lo kubectl https://storage.googleapis.com/kubernetes-release/release/v1.7.0/bin/linux/amd64/kubectl && chmod +x kubectl && sudo mv kubectl /usr/local/bin/
```
Comment (asymmetric) [inline travis-setup.sh]: Can we extract the kube version to a variable, here and below?

File: travis-setup.sh [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+# Fix the kubectl context, as it's often stale.
+minikube update-context
+
+# Hack for waiting for minikube to be ready.
+echo "sleeping..."
+sleep 3m
+
+echo "creating image"
+make TAG=testing image
```
Comment (asymmetric) [inline travis-setup.sh]: Idea for a next iteration: use [build stages](https://docs.travis-ci.com/user/build-stages/) to build the image in parallel, as the tests are being run.

File: travis-setup.sh (hunk 4 above)
Comment (LiliC) [reply travis-setup.sh]: Will test, it was needed when the config files were under root, but now that is moved to `USER`, we might not need it anymore.

File: travis-setup.sh (hunk 3 above)
Comment (LiliC) [reply travis-setup.sh]: Sure we can either do that, or if you prefer we can have it all in the `.travis.yml` file? I just separated as it was getting long and not that readable, but can move it back there to keep it all in one file. WDYT?

File: test/e2e/operator_test.go (hunk 2 above)
Comment (LiliC) [reply test/e2e/operator_test.go]: Another service is created again in the other test below, so we just want to make sure to remove this service before creating the other one.

File: Makefile (hunk 1 above)
Comment (LiliC) [reply Makefile]: It was not building the operator on Travis CI properly with the previous command, this seems to do the trick.

File: travis-setup.sh (hunk 7 above)
Comment (LiliC) [reply travis-setup.sh]: Cool, feel free to open issue for it, so we don't forget. Agreed also about bringing down the time of the tests, we can maybe make the issue generic `Improve tests performance` or something and note that as well.

File: travis-setup.sh (hunk 4 above)
Comment (LiliC) [reply travis-setup.sh]: We still need sudo:
```
E0906 13:50:32.697347    4238 start.go:158] Error updating cluster:  error creating file at /usr/local/bin/localkube: open /usr/local/bin/localkube: permission denied
```

**Comment by LiliC** [comment]
@asymmetric think I addressed/answered all the comments and tests are 🍏 . Can you PTALA, thanks!

File: travis-setup.sh (hunk 4 above)
Comment (asymmetric) [reply travis-setup.sh]: Is that because we move the binaries to `/usr/local/bin` though?

File: travis-setup.sh (hunk 3 above)
Comment (asymmetric) [reply travis-setup.sh]: Yeah I think it could be a good idea to move it back there, as it's only used in the context of Travis.

File: Makefile (hunk 1 above)
Comment (asymmetric) [reply Makefile]: I see. Is there any documentation we can point to to explain why they're there?

**Comment by asymmetric** [comment]
Addressed comments, including one that's probably hidden, so make sure to expand ;)

File: Makefile (hunk 1 above)
Comment (LiliC) [reply Makefile]: Which ones do you mean? The `go build` flags, are well documented already. And the variables think those are self-explanatory. But I think both are documented enough, not sure we need to leave a comment. :)

File: travis-setup.sh (hunk 4 above)
Comment (LiliC) [reply travis-setup.sh]: Got it from the minikube installation [instructions](https://github.com/kubernetes/minikube#linux). Not sure it matters if we do `sudo` or not, as it's just for tests setup.

**Comment by LiliC** [comment]
Done @asymmetric. Can you have another look, danke!

File: Makefile (hunk 1 above)
Comment (asymmetric) [reply Makefile]: If I would look at the Makefile, I wouldn't know why they work, and it seems they have been added "heuristically". So I think it would make sense to justify their presence.

TL;DR: If they're needed, we should explain why IMO.

File: travis-setup.sh (hunk 4 above)
Comment (asymmetric) [reply travis-setup.sh]: True, it doesn't really matter, but it's more hygienic IMO to only require privileges when they're strictly needed, as a general rule.

Comment (LiliC) [reply travis-setup.sh]: Good point, will try to get rid of it.
